src/controllerclientimpl.h

```diff
@@ -216,6 +220,7 @@ class ControllerClientImpl : public ControllerClient, public boost::enable_share
     ///
     /// overwrites the file if it already exists.
     virtual void _UploadFileToControllerViaForm(const std::string& file, const std::string& filename);
+    virtual void _UploadFileToControllerViaForm2(const std::string& file, const std::string& filename, const std::string& endpoint);
```

Comment:
There are only 2 places that use `_UploadFileToControllerViaForm`, please change that function and have the caller call with endpoint instead.